README.md
Distiller
=========

Distiller provides convenient auto-extraction of document key words
based on term-frequency/inverse-document-frequency (TF-IDF) and word
positioning.

Distiller handles all of the pre-processing details and produces final
statistic reports in JSON format.


Requirements
------------

Distiller uses the [Natural Language Toolkit](http://www.nltk.org/)

You will need to download a couple of NLTK packages:

    >>> import nltk
    >>> nltk.download()
    Downloader> d
    Download which package (l=list; x=cancel)?
        Identifier> maxent_treebank_pos_tagger
    Downloader> d
    Download which package (l=list; x=cancel)?
        Identifier> stopwords



Installation
------------

Installation using pip:

    $ pip install Distiller


Usage
-----

Typical usage from within the Python interpreter:

    >>> from Distiller.distiller import Distiller
    >>> distiller = Distiller(data, target, options)


Arguments
---------

### data

Path to file containing the document collection in JSON format.

    {
        'metadata': {
            'base_url': 'The document's source URL (if any)'
            },
        'documents': [
                {
                    'id': 'The document's unique identifier (if any)',
                    'body': 'The entire body of the document in a single text blob.',
                    'description': 'An optional brief description of this doc.'
                }, ...
            ]
    }

###target

Path where Distiller will output the following reports:

keywords: A list of words and the frequency with which they were detected as being keywords of documents.

bigrams: A list of word pairs and the frequency with which they were detected as being key pairs in documents.

trigrams: A list of word triples and the frequency with which they were detected as being key pairs in documents.

docmap: A mapping of document IDs to their respective keywords, n-grams, and other statistics.

keymap: A mapping of keywords to the documents they appear in.


###options

An optional dictionary containing document processing arguments in this format:

    {
        'normalize': True,          # normalize tokens during pre processing
        'stem': True,               # stems tokens during pre processing
        'lemmatize': False,         # lemmatize during pre processing
        'tfidf_cutoff': 0.001,      # cutoff value to use for term-freq/doc-freq score
        'pos_list': ['NN','NNP'],   # POS white list used to filter for candidates
        'black_list': []            # token list used to filter out from candidates
    }



CHANGES.txt
v0.1.2, 2014-08-13
    - Added logging and verbosity to public methods.
    - Internal code changes for processing pipeline and feature extraction.

v0.1.1, 2014-06-20 
    - Package reorganization
    - Test data added.
    - Basic tests added.

v0.1, 2014-06-19 
    - Initial Release.


Distiller/test/test_distiller.py
from dircache import listdir
import os
import unittest
from Distiller.distiller import Distiller

test_data = {
        'metadata': {
            'base_url': 'http://github.com/franciscocanas/Distiller/'
        },
        'documents': [
            {
                'id': 1,
                'body': "Far far away, behind the word mountains, far from the countries "
                        "Vokalia and Consonantia, there live the blind texts. Separated they "
                        "live in Bookmarksgrove right at the coast of the Semantics, a large "
                        "language ocean. A small river named Duden flows by their place and "
                        "supplies it with the necessary regelialia. It is a paradisematic country, "
                        "in which roasted parts of sentences fly into your mouth. Even the "
                        "all-powerful Pointing has no control about the blind texts it is an "
                        "almost unorthographic life One day however a small line of blind text "
                        "by the name of Lorem Ipsum decided to leave for the far World of Grammar. "
                        "The Big Oxmox advised her not to do so, because there were thousands of "
                        "bad Commas, wild Question Marks and devious Semikoli, but the Little Blind Text did not "
                        "listen. She packed her seven versalia, put her initial into the belt and made herself "
                        "on the way."
            },
            {
                'id': 2,
                'body': "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed "
                        "in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted "
                        "his head a little he could see his brown belly, slightly domed and divided by arches "
                        "into stiff sections. The bedding was hardly able to cover it and seemed ready to "
                        "slide off any moment. His many legs, pitifully thin compared with the size of the "
                        "rest of him, waved about helplessly as he looked. \"What's happened to me?\\"
            },
            {
                'id': 3,
                'body': "A wonderful serenity has taken possession of my entire soul, like these sweet "
                        "mornings of spring which I enjoy with my whole heart. I am alone, and feel the "
                        "charm of existence in this spot, which was created for the bliss of souls like "
                        "mine. I am so happy, my dear friend, so absorbed in the exquisite sense of mere "
                        "tranquil existence, that I neglect my talents. I should be incapable of drawing "
                        "a single stroke at the present moment; and yet I feel that I never was a greater "
                        "artist than now. When, while the lovely valley teems with vapour around me, and the "
                        "meridian sun strikes the upper surface of the impenetrable foliage of my trees, "
                        "and but a few stray gleams steal into the inner sanctuary, I throw myself down "
                        "among the tall grass by the trickling stream"
            }
        ]
    }

data = 'data/data.json'
result = 'data/tests/'

nlp_args = {
    'normalize': True,
    'stem': False,
    'lemmatize': False,
    'tfidf_cutoff': 0.0001,
    'pos_list': ['NN','NP', 'JJ'],
    'black_list': []
}

def clean_folder(path):
    """
    Remove any existing files or folders at path.
    """
    if os.path.exists(path):
        for fileName in listdir(path):
            os.remove(path + '/' + fileName)
        os.rmdir(path)

class TestDistiller(unittest.TestCase):
    """
    Basic testing for Distiller class.
    """

    def test_DistillerOutputFiles(self):
        """
        Runs Distiller on test data and checks that output files are produced.
        """
        clean_folder(result)
        Distiller(data, result, nlp_args, verbosity=3)
        self.assertTrue(os.path.exists(result + 'keymap.json'))
        self.assertTrue(os.path.exists(result + 'docmap.json'))
        self.assertTrue(os.path.exists(result + 'keywords.json'))
        self.assertTrue(os.path.exists(result + 'bigrams.json'))
        self.assertTrue(os.path.exists(result + 'trigrams.json'))

    def test_DistillerDocs(self):
        """
        Runs on test data and checks that all stats are collected.
        """
        clean_folder(result)
        d = Distiller(data, result, nlp_args, verbosity=3)
        self.assertTrue(d.processed_documents)
        self.assertTrue(d.statistics)
        self.assertTrue(d.statistics['keywords'])
        self.assertTrue(d.statistics['bigrams'])
        self.assertTrue(d.statistics['trigrams'])







Distiller/features/tf_idf.py
import math
import nltk

__author__ = 'mailfrancisco@gmail.com'


class TF_IDF():
    def __init__(self, documents):
        self.idf_cache = {}
        self.documents = documents
        self.docs_number = len(documents)

    def compute(self, candidates, document):
        """
        For each token used in candidates list:
        Compute the final tf-idf score. The product of tf * idf.
        Returns a list of tuples: (word, tf-idf score)
        Sorted by tf-idf score.
        """
        tfs = self.compute_tf(candidates, document)
        idfs = self.compute_idf(candidates)
        return sorted(map(lambda x, y: (x[0], x[1] * y[1]), tfs, idfs), key=lambda item: item[1], reverse=True)

    def compute_tf(self, candidates, document):
        """
        Compute the term frequency: How many times this word appears
        in this document.
        Returns a list of tuples: (word, tf score)
        """
        dist = nltk.FreqDist(document)
        score_tf = lambda w, doc: dist[w.lower()] / float(len(doc))
        return [(word, score_tf(word, document)) for word in sorted(candidates)]

    def compute_idf(self, candidates):
        """
        Compute the inverse document frequency: A score of how uncommon a word
        is among all the documents.
        log(total_number_of_docs/1 + number_of_docs_tag_appears_in)
        Returns a list of tuples: (word, idf score)
        """
        idfs = []
        score_idf = lambda w: sum([1 for d in self.documents if w.lower() in d])

        for word in set(sorted(candidates)):
            if not self.idf_cache.has_key(word):
                idf = math.log(self.docs_number / 1.0 + score_idf(word))
                self.idf_cache[word] = idf
            else:
                idf = self.idf_cache[word]
            idfs.append((word, idf))
        return idfs




Distiller/preprocessing/pipeline.py
from nltk import RegexpTokenizer
from nltk.corpus import stopwords
import re
import nltk

__author__ = 'fcanas'


class Pipeline():
    """
    Responsible for pre-processing any given document by performing the following set of steps:
    1) Divide content into tokens.
    2) POS tag tokens.
    3) Filter out only potential candidates.
    3) Normalize
    4) Filter out useless tokens.
    5) Stem.
    6) Lemmatize
    """

    def __init__(self, pos_tag=True, black_list=None, pos_list=None, normalize=True, stem=True, lemmatize=True):
        if not pos_list:
            pos_list = ['NN', 'JJ', 'NNP']
        if not black_list:
            black_list = []
        self.pos_tag = pos_tag
        self.black_list = black_list
        self.pos_list = pos_list
        self.normalize = normalize
        self.stem = stem
        self.lemmatize = lemmatize

    def pre_process(self, text):
        """
        Input: "Blob of text..."
        Output: [(token, tag), ...]
        """
        processed = []
        tokens = self.tokenize(text)
        if self.pos_tag:
            tokens = self.pos_tag_tokens(tokens)

        for token in tokens:
            if not self.filter_by_pos(token) or not self.filter_by_pattern(token):
                continue
            if self.normalize:
                token = self.normalize_text(token)
            if self.stem:
                token = self.stem_token(token)
            if self.lemmatize:
                token = self.lemmatize_token(token)
            processed.append(token)
        return processed

    @staticmethod
    def tokenize(text):
        """
        Input: "Body of text...:
        Output: [word, ...] list of tokenized words matching regex '\w+'
        """
        tokenizer = RegexpTokenizer(r'\w+')
        tokens = tokenizer.tokenize(text)
        return tokens

    @staticmethod
    def pos_tag_tokens(tokens):
        """
        Use nltk's pos taggers on the given tokens.
        Input: [token1, token2, ...]
        Output: [(token1, tag1), (token2, tag2), ...]
        """
        return nltk.pos_tag(tokens)

    def filter_by_pos(self, token):
        """
        Input: (word, pos)
        Output: True if pos is noun, proper noun, or adj. False otherwise.
        """
        return token[1] in self.pos_list

    @staticmethod
    def normalize_text(token):
        """
        Input: (word, pos)
        Output: (word, pos) with lower case word.
        """
        return token[0].lower(), token[1]

    @staticmethod
    def stem_token(token):
        """
        Input: (word, pos)
        Output: (stem of word, pos)
        """
        porter = nltk.PorterStemmer()
        return porter.stem(token[0]), token[1]

    @staticmethod
    def lemmatize_token(token):
        """
        Input: (word, pos)
        Output: (Lemmatized word, pos)
        """
        wnl = nltk.WordNetLemmatizer()
        return wnl.lemmatize(token[0]), token[1]

    def filter_by_pattern(self, token):
        """
        Filters by removing tokens with words that are numbers, or in the black list or in the stop words list.

        Input: (word, pos), [blacklist_word, ...]
        Output: False if word is any of the above. True otherwise.
        """
        nums = re.compile("^[0-9]*$")
        filtered = token[0].lower() not in map(unicode, self.black_list) and \
                   token[0].lower() not in stopwords.words('english') and not \
                       nums.match(token[0])
        return filtered


distiller/src/features/Positioning.py

__author__ = 'mailfrancisco@gmail.com'

class Positioning():

    def __init__(self):
        pass

    def compute_position_score(self, tokens, document):
        """
        Compute the positioning score for each of the tokens in this
        document and return a map of token => score.
        """
        self.scores = {}
        for token in tokens:
            if not token in self.scores.keys():
                self.scores[token] = self.positioning_score(token, document)
        return self.scores

    def first_occurrence(self, token, body):
        """
        Return the index of the first occurrence of token in
        the body of tokens.
        """
        try:
            index = body.index(token)
        except ValueError:
            index = len(body)
        return index

    def positioning_score(self, token, body):
        """
        Return a score based on the position of token in a
        body of tokens, normalized by length.
        """
        return 1.0 + (1.0 - float(self.first_occurrence(token, body)) / float(len(body)))

distiller/src/features/Collocations.py
import nltk
from nltk.collocations import *

__author__ = 'mailfrancisco@gmail.com'

class Collocations():

    def __init__(self):
        pass

    def find_ngrams(self, n=2, body=None, top=10):
        """
        Input
        n:  n-grams!
        body: A list containing a sequence of words from a text.
        top: How many n-grams to return.

        Output:
        A list of the top 'top' n-grams in body.
        """
        measures = {
            2: nltk.collocations.BigramAssocMeasures,
            3: nltk.collocations.TrigramAssocMeasures
        }

        finder = {
            2: BigramCollocationFinder,
            3: TrigramCollocationFinder
        }


        if not body:
            return []

        m = measures[n]()
        result = finder[n].from_words(body)
        return result.nbest(m.pmi, top)


Distiller/distiller.py
import os
import json
import logging
import nltk

from features.Collocations import Collocations
from features.Positioning import Positioning
from features.tf_idf import TF_IDF
from preprocessing.pipeline import Pipeline


__author__ = 'fcanas'


class Distiller():
    """
    Responsible for extracting features from a set of documents. Expects documents specified in JSON format:

    {
    metadata : {
        base_url : "..."
        }
    documents : [{
        id: INT,
        description: "..." # optional
        body: "...",
        }, ...],
    }


    The documents will be processed and kept in processed_documents dic, which will have this format:

    {
    id => {
            id : INT,

            tokenized_body : [token1, token2, ...] # where tokens are lowerecase unicode words that form text body

            # the resulting list from calling pre_process_pipeline()
            processed_body : [(processed_token1, POS1), (processed_token2, POS2),...],

            candidates: [candidate1, candidate2, ...], # likely candidate words are determined by pipeline

            freq_distribution:  {token1: frequency2, token2: frequency2} # frequencies for all tokens in body
        }
    }

    nlp_args:

    A dictionary of arguments used by the pre-processing pipeline. Takes the following form:

    {
        normalize: Boolean,                 # normalize tokens during pre processing
        stem: Boolean,                      # stems tokens during pre processing
        lemmatize: Boolean,                 # lemmatize during pre processing
        tfidf_cutoff: Float,                # cutoff value to use for term-freq/doc-freq score
        pos_list: [STRING,...],             # POS white list used to filter for candidates
        black_list: [token1, token2, ...]   # token list used to filter out from candidates
    }

    """

    default_args = {
        'normalize': True,  # normalize tokens during pre processing
        'stem': True,  # stems tokens during pre processing
        'lemmatize': False,  # lemmatize during pre processing
        'tfidf_cutoff': 0.001,  # cutoff value to use for term-freq/doc-freq score
        'pos_list': ['NN', 'NNP'],  # POS white list used to filter for candidates
        'black_list': []  # token list used to filter out from candidates
    }

    def __init__(self, document_file, target_path, nlp_args=default_args, verbosity=2):
        """
        Initialize Distiller for specified document file.
        """
        self.processed_documents = {}
        self.statistics = {}
        self.collocations = Collocations()
        self.positioning = Positioning()
        self.path = make_path(target_path)
        logging.basicConfig(format='%(asctime)s %(levelname)s: %(message)s',
                            level=self.get_logging_level(verbosity))

        with open(document_file) as d:
            self.jdata = json.load(d)
        d.close()

        self.initialize_arguments(nlp_args)
        self.process_documents()
        self.tfidf = TF_IDF(self.processed_doc_bodies)
        self.extract_features()
        self.compile()
        self.export()

    def initialize_arguments(self, nlp_args):
        """
        Read args from json metadata and initialize Distiller arguments.
        """
        logging.info("initializing Distiller")
        self.base_url = self.jdata['metadata']['base_url']
        self.documents = self.jdata['documents']
        self.normalize = nlp_args.get('normalize', self.default_args['normalize'])
        self.stem = nlp_args.get('stem', self.default_args['stem'])
        self.lemmatize = nlp_args.get('lemmatize', self.default_args['lemmatize'])
        self.pos_list = nlp_args.get('pos_list', self.default_args['pos_list'])
        self.tfidf_cutoff = nlp_args.get('tfidf_cutoff', self.default_args['tfidf_cutoff'])
        self.black_list = nlp_args.get('black_list', self.default_args['black_list'])

    def process_documents(self):
        """
        Run documents from json through pre-processing.
        """
        pipeline = Pipeline(black_list=self.black_list,
                            pos_list=self.pos_list,
                            normalize=self.normalize,
                            stem=self.stem,
                            lemmatize=self.lemmatize)

        for document in self.documents:
            logging.info("processing document {0}".format(document['id']))

            doc = {
                'id': document['id'],
                'url': self.base_url.format(int(document['id'])),
                'tokenized_body': map(unicode.lower, nltk.word_tokenize(document['body'])),
                'processed_tokens': pipeline.pre_process(text=document['body']),
                'description': document.get('description', '')
            }

            if not doc['processed_tokens']:
                doc['candidates'] = []
            else:
                doc['candidates'] = list(set(zip(*doc['processed_tokens'])[0]))
            doc['freq_distribution'] = nltk.FreqDist(doc['tokenized_body'])
            self.processed_documents[doc['id']] = doc

        self.processed_doc_bodies = [document['processed_tokens'] for document in self.processed_documents.values()]

    def extract_features(self):
        """
        Extract the features for each pre-processed document, given the entire body of docs.
        """
        for document in self.processed_documents.values():
            logging.info("computing statistics for {0}".format(document['id']))
            document['tfidf'] = self.tfidf.compute(
                document['candidates'],
                document['tokenized_body'])

            document['positioning'] = self.positioning.compute_position_score(document['candidates'],
                                                                              document['tokenized_body'])

            document['keywords'] = self.extract_keywords(document['tfidf'],
                                                         document['positioning'],
                                                         lower_cutoff=self.tfidf_cutoff)

            document['bigrams'] = self.collocations.find_ngrams(2, document['processed_tokens'])

            document['trigrams'] = self.collocations.find_ngrams(2, document['processed_tokens'])

    @staticmethod
    def extract_keywords(tf_idf_scores, positioning_scores, lower_cutoff=0.0001):
        """
        Input: sorted list of tf-idf candidate scores
               hash of position scores.

        output: list of (candidate, score=tf-idf * positioning)
        """
        keywords = []
        for candidate in tf_idf_scores:
            if candidate[1] > lower_cutoff:
                score = candidate[1] * 2 * positioning_scores[candidate[0]]
                keywords.append((candidate[0], score))
        return keywords

    def compile(self):
        """
        Compile the statistics of all extracted features.
        """
        self.compile_statistic('keywords', lambda x: x[0], nltk.FreqDist)
        self.compile_statistic('bigrams', lambda x: ' '.join(map(lambda y: y[0], x)), nltk.FreqDist)
        self.compile_statistic('trigrams', lambda x: ' '.join(map(lambda y: y[0], x)), nltk.FreqDist)
        self.compile_collections()

    def export(self):
        """
        Write all of the stats and processed documents out to the target path.
        """
        logging.info("exporting statistics to {0}".format(self.path))
        for key, value in self.statistics.items():
            export_to_file(self.path, key, value)

        export_to_file(self.path, 'keymap', self.keymap)
        export_to_file(self.path, 'docmap', self.docmap)

    def compile_statistic(self, stat, transformer=lambda x: x, compiler=lambda x: x):
        """
        Creates a json output file for the given stat and set of bugs.
        """
        stats = []
        for doc in self.processed_documents.values():
            for item in doc[stat]:
                stats.append(transformer(item))
        self.statistics[stat] = compiler(stats)

    def compile_collections(self):
        """
        Stores the collections:
        (keyword => [BZ id, ...])
        (BZ id => Bug)
        """
        self.keymap = {}
        self.docmap = {}
        logging.info("storing document and keyword collections to {0}".format(self.path))
        for doc in self.processed_documents.values():
            self.docmap[doc['id']] = doc
            for word in doc['keywords']:
                if not word[0] in self.keymap:
                    self.keymap[word[0]] = []
                self.keymap[word[0]].append(str(doc['id']))

    @staticmethod
    def get_logging_level(verbosity):
        """
        Return a logging level based on verbosity argument {0,1,2}.
        """
        if verbosity < 1:
            return logging.ERROR
        if verbosity == 1:
            return logging.WARNING
        if verbosity == 2:
            return logging.INFO
        if verbosity > 2:
            return logging.DEBUG


def export_to_file(path, stat, col):
    """
    Export the given stat to file in the given path.
    """
    with open(path + '/' + stat + '.json', 'w') as outfile:
        json.dump(col, outfile)
    outfile.close()


def make_path(path):
    """
    Ensures the target path for stat reports is created.
    """
    if not path.endswith('/'):
        path += '/'

    if not os.path.exists(os.path.dirname(path)):
        os.makedirs(os.path.dirname(path))

    return path
